### PR TITLE
AB#32151: Add ExtractionProcedure API Controller and Service Methods

### PR DESCRIPTION
**Pull request by shakisparki**
## Overview

Add an API Controller for Extraction Procedure - and create the required service methods alongside.

Note ExtractionProcedure is now a SnomedTerm and will need to be filtered from that table

Some bug fixes.

### CODE REVIEW

**Review by Flemingjp — APPROVED** [review]

File: src/Directory/Biobanks.Web/ApiControllers/ExtractionProcedureController.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                .Result
+            )
+            .ToList();
+        }
+
+        [HttpDelete]
+        [Route("{id}")]
+        public async Task<IHttpActionResult> Delete(string id)
+        {
+            var model = (await _biobankReadService.ListExtractionProceduresAsync()).Where(x => x.Id == id).First();
```
Comment (Flemingjp) [inline src/Directory/Biobanks.Web/ApiControllers/ExtractionProcedureController.cs:50]: Better of creating a `GetExtractionProcedure` service method? That way you can do the filtering on database

**Review by Flemingjp — CHANGES_REQUESTED** [review]

**Review by Flemingjp — DISMISSED** [review]

File: src/Directory/Biobanks.Web/ApiControllers/DiseaseStatusController.cs [hunk 2]
```diff
@@ -79,14 +79,14 @@ public async Task<IHttpActionResult> Put(string id, OntologyTermModel model)
             //If this description is valid, it already exists
             if (await _biobankReadService.ValidOntologyTermDescriptionAsync(id, model.Description))
             {
-                ModelState.AddModelError("Description", "That description is already in use by another disease status. Disease status descriptions must be unique.");
+                ModelState.AddModelError("Description", "That description is already in use by another disease status or extraction procedure. Disease status descriptions must be unique.");
```
Comment (beforan) [inline src/Directory/Biobanks.Web/ApiControllers/DiseaseStatusController.cs:82]: I feel like this message is confusing everywhere it appears. In this particular case, the first clause mentions both ontology term types, but the second only mentions disease status, and if they're editing disease statuses mentioning extraction procedure out of the blue is just gonna confuse them.

I think something like "That description is already in use. Descriptions must be unique across all ontology terms." is less confusing but still accurate.

File: src/Directory/Biobanks.Web/ApiControllers/DiseaseStatusController.cs [hunk 3]
```diff
@@ -115,10 +115,14 @@ public async Task<IHttpActionResult> Put(string id, OntologyTermModel model)
         [Route("")]
         public async Task<IHttpActionResult> Post(OntologyTermModel model)
         {
+            //if ontology term id is in use by another ontology term
+            if ((await _biobankReadService.ListOntologyTermsAsync()).Any(x => x.Id == model.OntologyTermId))
+                ModelState.AddModelError("OntologyTermId", "This Id is currently in use by another Disease status or extraction procedure. Ontology Term Ids must be unique");
```
Comment (beforan) [inline src/Directory/Biobanks.Web/ApiControllers/DiseaseStatusController.cs:120]: I would word this the same as the description message, but with "ID" instead.

File: src/Directory/Biobanks.Web/ApiControllers/ExtractionProcedureController.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+using System.Web.Http;
+using Biobanks.Web.Models.Shared;
+using Biobanks.Web.Models.ADAC;
+using System.Collections;
+using Biobanks.Entities.Shared.ReferenceData;
+using Biobanks.Web.Filters;
+
+namespace Biobanks.Web.ApiControllers
+{
+    //[UserApiAuthorize(Roles = "ADAC")]
```
Comment (beforan) [inline src/Directory/Biobanks.Web/ApiControllers/ExtractionProcedureController.cs:13]: why is this commented out? is that a mistake or should it be removed?

File: src/Directory/Biobanks.Web/ApiControllers/ExtractionProcedureController.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        }
+
+        [HttpPut]
+        [Route("{id}")]
+        public async Task<IHttpActionResult> Put(string id, OntologyTermModel model)
+        {
+            //If this description is valid, it already exists
+            if (await _biobankReadService.ValidOntologyTermDescriptionAsync(id, model.Description))
+            {
+                ModelState.AddModelError("Description", "That description is already in use by another disease status or extraction procedure. Extraction procedure descriptions must be unique.");
```
Comment (beforan) [inline src/Directory/Biobanks.Web/ApiControllers/ExtractionProcedureController.cs:83]: Simplify this / make it consistent with elsewhere

**Review by beforan — CHANGES_REQUESTED** [review]

**Comment by blargly** [comment]
No longer needed

**Pull request closed without merging**